File: convex/lib/normalize.ts
/**
 * Canonical card-name normalization. Casing/punctuation/diacritics-insensitive.
 *
 * IMPORTANT: agents/resolver.py (Phase 2) MUST replicate this algorithm exactly,
 * or exact-match lookups will silently miss. Algorithm:
 *   1. Unicode NFKD normalize, strip combining marks (diacritics): "Æther"->"aether-ish"
 *   2. lowercase
 *   3. drop apostrophes entirely (no space): "Urza's" -> "urzas"
 *   4. replace every other non [a-z0-9] run with a single space
 *   5. trim + collapse whitespace
 *
 * Examples:
 *   "Jace, the Mind Sculptor" -> "jace the mind sculptor"
 *   "Fire // Ice"             -> "fire ice"
 *   "Lim-Dûl's Vault"         -> "lim duls vault"
 */
export function normalizeName(name: string): string {
  return name
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/['\u2019]/g, "")
    .replace(/[^a-z0-9]+/g, " ")
    .trim()
    .replace(/\s+/g, " ");
}

File: agents/core/normalize.py
"""Canonical card-name normalization — Python mirror of convex/lib/normalize.ts.

These two implementations MUST stay byte-for-byte equivalent, or exact-match
card lookups (Python side) will miss rows written by the TS ingester.
"""

from __future__ import annotations

import re
import unicodedata

_APOSTROPHES = re.compile(r"['\u2019]")
_NON_ALNUM = re.compile(r"[^a-z0-9]+")
_WHITESPACE = re.compile(r"\s+")


def normalize_name(name: str) -> str:
    """Lowercase, strip diacritics/apostrophes/punctuation, collapse whitespace."""
    decomposed = unicodedata.normalize("NFKD", name)
    without_marks = "".join(c for c in decomposed if not unicodedata.combining(c))
    lowered = without_marks.lower()
    no_apostrophes = _APOSTROPHES.sub("", lowered)
    spaced = _NON_ALNUM.sub(" ", no_apostrophes)
    return _WHITESPACE.sub(" ", spaced).strip()
